File: Recursion/13. Sudoku Solver/sudoku_solver.cpp
// https://leetcode.com/problems/sudoku-solver

class Solution {
public:

    bool isValid(int row, int col, vector<vector<char>>& board, char c) {
        for(int i = 0; i<9; i++) {
            //check row
            if(board[row][i] == c) return false;
            //check col
            if(board[i][col] == c) return false;
            //check box
            if(board[3*(row/3) + (i/3)][3*(col/3) + (i%3)] == c) return false;
        }
        return true;
    }

    bool helper(vector<vector<char>>& board) {
        for(int i = 0; i<9; i++) {
            for(int j = 0; j<9; j++) {
                if(board[i][j] == '.') {
                    for(char c = '1'; c <= '9'; c++) {
                        if(isValid(i, j, board, c)) {
                            board[i][j] = c;
                            if(helper(board)) return true;
                            board[i][j] = '.';
                        }
                    }
                    return false;
                }
            }
        }
        return true;
    }

    void solveSudoku(vector<vector<char>>& board) {
        helper(board);
    }
};

File: Recursion/13. Sudoku Solver/sudoku_solver.py
# https://leetcode.com/problems/sudoku-solver

# Actual problem: Place numbers(1 to 9) at 'E' places
# Sub-problem: Place numbers(1 to 9) at 1,2,....E-1,E places

class Solution:
    def solveSudoku(self, board: List[List[str]]) -> None:
        self.placeNumberOnBoard(board)
    def placeNumberOnBoard(self,board: List[List[str]])->bool:
        for row in range(9):
            for col in range(9):
                if board[row][col]==".":
                    for c in range(1,10):
                        c=str(c)
                        if self.isValid(row,col,board,c):
                            board[row][col]=c
                            if self.placeNumberOnBoard(board):
                                return True
                            board[row][col]="."
                    return False
        return True
    
    def isValid(self,row:int,col:int,board: List[List[str]], c:str)->bool:
        for i in range(9):
            if board[row][i]==c:
                return False
            if board[i][col]==c:
                return False
            if board[3*(row//3)+i//3][3*(col//3)+i%3]==c:
                return False
        return True